csvcubed/csvcubed/readers/columnschema.py

```diff
@@ -0,0 +1,447 @@
+    from_existing: Optional[str] = None
+    definition_uri: Optional[str] = None
+    scaling_factor: Optional[float] = None
+    si_scaling_factor: Optional[float] = None
+    quantity_kind: Optional[str] = None
+
+
+@dataclass
+class Units(SchemaBaseClass):
+    values: Union[bool, List[UnitType]] = True
+
+    def map_to_qb_multi_units(self, data: PandasDataTypes) -> QbMultiUnits:
+        if isinstance(self.values, bool) and self.values is True:
+            return QbMultiUnits.new_units_from_data(data)
+
+        elif isinstance(self.values, list):
+
+            units = []
+            for unit in self.values:
+                if not isinstance(unit, UnitType):
+                    raise ValueError(f"Unexpected unit value: {unit}")
+
+                if any([unit.label, unit.description, unit.definition_uri]):
+                    # NewUnit
+                    units.append(_map_unit(unit))
+
+                elif unit.from_existing:
+                    units.append(ExistingQbUnit(unit_uri=self.values.from_existing))
+
+                else:
+                    raise ValueError(f"Unexpected unit value: {unit}")
+
+            return QbMultiUnits(units)
+
+        raise ValueError(f"Unhandled 'Units' value: {self}")
+
+
+@dataclass
+class MeasureType(SchemaBaseClass):
+    label: Optional[str] = None
+    description: Optional[str] = None
+    from_existing: Optional[str] = None
+    definition_uri: Optional[str] = None
+
+
+@dataclass
+class Measures(SchemaBaseClass):
+    values: Union[bool, List[MeasureType]] = True
+
+    def map_to_multi_measure_dimension(
+        self, data: PandasDataTypes
+    ) -> QbMultiMeasureDimension:
+        # When values is a single bool True then create new Measures from the csv column data
+        if self.values is True:
+            return QbMultiMeasureDimension.new_measures_from_data(data)
+
+        elif isinstance(self.values, list):
+            new_measures = []
+            for new_measure in self.values:
+                if not isinstance(new_measure, MeasureType):
+                    raise ValueError(f"Unexpected measure: {new_measure}")
+
+                if any([new_measure.label, new_measure.description, new_measure.definition_uri]):
+                    # NewQBMeasure
+                    new_measures.append(_map_measure(new_measure))
+                else:
+                    new_measures.append(QbMultiMeasureDimension(new_measure))
+
+            return QbMultiMeasureDimension(new_measures)
+
+        else:
+            raise ValueError(f"Unexpected values type: {self.values}")
+
+
+@dataclass
+class ObservationValue(SchemaBaseClass):
+    datatype: Optional[str] = None
+    unit: Union[None, str, Units] = None
+    measure: Union[None, str, Measures] = None
+
+    def map_to_qb_observation(self) -> QbObservationValue:
+        unit = None
+        if isinstance(self.unit, str):
+            unit = ExistingQbUnit(unit_uri=self.unit)
+
+        elif isinstance(self.unit, Units):
```

Comment:
Again, this needs to be `UnitType` here.

Reply:
fixed